### DISCUSSION TITLE
vcpkg support

### DISCUSSION THREAD

**Discussion by webfolderio** [taocpp/json#50] [Ideas]
Hi,

Do you consider to support [vcpkg](https://github.com/Microsoft/vcpkg)?

**Comment by d-frey** [comment]
We (Colin and me) are not using Windows, hence we usually don't support it directly from our side. Someone (@uilianries / @wravery ?) needs to send a PR to them. I'm not sure if this is reasonable pre-1.0.0, but large parts seem quite stable so it could make sense.

**Comment by uilianries** [comment]
Since VCPKG is a Conan's rival I can't help with it :joy: Joking ...

I don't have experience packaging projects for vcpkg, but I think it's a good idea, as it is very popular and it will help to distribute this project over Windows.

@webfolderio Do you know how to create a package there? If not, I would like to try, to understand the packaging process.

**Comment by wravery** [comment]
@uilianries If you have vcpkg installed/bootstrapped, you can start a new package with `vcpkg create <package> <archive URL>`. Once the package is initialized, you can open its files with `vcpkg edit <package>`, which will launch whatever editor you have in the `EDITOR` environment variable (or fall back to VS Code if it can find it).

It starts with a path/URL to an archive file rather than a GitHub repo/commit, but you can replace the `vcpkg_download_distfile` call in `portfile.cmake` with a call to `vcpkg_from_github`, e.g.

```cmake
vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO taocpp/pegtl
    REF c5d158ffab4baa0e1c3f6d54c0684f05a2412f74
    SHA512 efd5d36ca14c27948005cd034e72e24855f6046141c316a1d4b7186af9753fd035dab8303dc86e2f2de1b185f5fa3f5d678672892b1fc4ab69e827492a94e897
    HEAD_REF master
)
```

BTW, that's from the `3.0.0-pre` update for `pegtl` I recently landed (with `2.8.0` moved to `pegtl-2`). Vcpkg cares about the version in the `CONTROL` file being a string match rather than doing a semver comparison to decide whether or not a package is upgradeable. So adding a `-pre` suffix is good enough to distinguish it from an eventual `1.0.0` release. Similarly, if you want to rev the package without changing the version string number, the convention is to just add `-*`, e.g. `-1`, `-2`, etc.

**Comment by wravery** [comment]
Once you have a draft of the package's `CONTROL` file and `portfile.cmake`, running `vcpkg build <package>` will attempt to build it without actually installing it. Header-only libraries with pre-existing CMake build systems are pretty quick and easy, but if it encounters errors in the CMake config or the build/install step it should give you error messages pointing to log files in the build tree. Then you just iterate until it builds clean.

Oh, and since it includes the `SHA512` hash of the commit, I usually find that when I want to rev the `REF` I have to build it once and fail to get the error message with the actual hash, then paste that in and proceed to build the updated package.

**Comment by uilianries** [comment]
@wravery Thanks for your excellent explanation!

**Comment by webfolderio** [comment]
Fixed by @myd7349

https://github.com/Microsoft/vcpkg/pull/6253

**Comment by ColinH** [comment]
Thanks!

**Comment by uilianries** [comment]
Amazing!
